File: analyze_challenge_actionability.js
const fs = require('fs');

// Read the SQL file
const sqlContent = fs.readFileSync('ascevo/supabase/lessons_all_pillars.sql', 'utf8');

// Extract all lessons with their challenge cards
const lessonRegex = /INSERT INTO lessons \(id, unit_id, title, display_order, card_concept, card_example, card_mistake, card_science, card_challenge\) VALUES\s*\n\('([^']+)','([^']+)','([^']+)',(\d+),\s*'((?:[^']|'')+)',\s*'((?:[^']|'')+)',\s*'((?:[^']|'')+)',\s*'((?:[^']|'')+)',\s*'((?:[^']|'')+)'\)/gs;

const lessons = [];
let match;

while ((match = lessonRegex.exec(sqlContent)) !== null) {
  const [_, id, unit_id, title, display_order, concept, example, mistake, science, challenge] = match;
  lessons.push({
    id,
    title,
    challenge: challenge.replace(/''/g, "'") // Unescape single quotes
  });
}

console.log(`\n========================================`);
console.log(`CHALLENGE ACTIONABILITY REVIEW`);
console.log(`Total lessons analyzed: ${lessons.length}`);
console.log(`========================================\n`);

// Criteria to check
const issues = [];

lessons.forEach((lesson, index) => {
  const challenge = lesson.challenge;
  const lessonNum = index + 1;
  
  // Check 1: Time-bound (should mention "today", "right now", "next time", specific duration)
  const timeIndicators = /\b(today|right now|now|next time|tonight|this week|tomorrow|in \d+ (minute|second|hour|day)s?|set a timer|for \d+)\b/i;
  const hasTimebound = timeIndicators.test(challenge);
  
  // Check 2: Specific action (should have clear verb and action)
  const actionVerbs = /\b(record|write|say|do|set|try|answer|practice|count|notice|ask|start|stop|pick|choose|call|text|send|take|make|create|list|identify|speak|read|watch|listen)\b/i;
  const hasSpecificAction = actionVerbs.test(challenge);
  
  // Check 3: Length check (should be 50-75 words ideally)
  const wordCount = challenge.split(/\s+/).length;
  const isAppropriateLength = wordCount >= 30 && wordCount <= 100;
  
  // Check 4: Success criteria (should include clear outcome or "notice")
  const successIndicators = /\b(notice|count|write down|compare|measure|track|record|feel|observe|check|see if|that's|you'll know)\b/i;
  const hasSuccessCriteria = successIndicators.test(challenge);
  
  // Check 5: Under 10 minutes (explicit or implicit)
  const quicklyCompletable = /\b(5 minute|10 minute|2 minute|3 minute|one minute|60 second|30 second|quick|brief|short)\b/i;
  const seemsQuick = quicklyCompletable.test(challenge) || wordCount < 80;
  
  // Identify issues
  const lessonIssues = [];
  
  if (!hasTimebound) {
    lessonIssues.push('❌ NOT TIME-BOUND: Missing immediate trigger (today, right now, next time, specific duration)');
  }
  
  if (!hasSpecificAction) {
    lessonIssues.push('❌ NOT SPECIFIC: Missing clear action verb or specific task');
  }
  
  if (!hasSuccessCriteria) {
    lessonIssues.push('❌ UNCLEAR SUCCESS: Missing clear outcome or "notice what happens" prompt');
  }
  
  if (!seemsQuick) {
    lessonIssues.push('⚠️  POSSIBLY TOO LONG: Challenge may take >10 minutes');
  }
  
  if (wordCount > 100) {
    lessonIssues.push(`⚠️  TOO WORDY: ${wordCount} words (should be 50-75)`);
  }
  
  if (wordCount < 30) {
    lessonIssues.push(`⚠️  TOO SHORT: ${wordCount} words (should be 50-75)`);
  }
  
  if (lessonIssues.length > 0) {
    issues.push({
      lessonNum,
      id: lesson.id,
      title: lesson.title,
      challenge,
      issues: lessonIssues,
      wordCount
    });
  }
});

// Report results
if (issues.length === 0) {
  console.log('✅ ALL CHALLENGES PASS ACTIONABILITY CRITERIA!\n');
} else {
  console.log(`⚠️  FOUND ${issues.length} CHALLENGES WITH POTENTIAL ISSUES:\n`);
  
  issues.forEach(issue => {
    console.log(`\n${'='.repeat(80)}`);
    console.log(`Lesson #${issue.lessonNum}: ${issue.title}`);
    console.log(`ID: ${issue.id}`);
    console.log(`Word count: ${issue.wordCount}`);
    console.log(`\nISSUES:`);
    issue.issues.forEach(i => console.log(`  ${i}`));
    console.log(`\nCHALLENGE TEXT:`);
    console.log(`"${issue.challenge}"`);
  });
  
  console.log(`\n\n${'='.repeat(80)}`);
  console.log(`SUMMARY`);
  console.log(`${'='.repeat(80)}`);
  console.log(`Total lessons: ${lessons.length}`);
  console.log(`Lessons with issues: ${issues.length}`);
  console.log(`Pass rate: ${((lessons.length - issues.length) / lessons.length * 100).toFixed(1)}%`);
  
  // Count issue types
  const issueTypes = {
    notTimebound: 0,
    notSpecific: 0,
    unclearSuccess: 0,
    tooLong: 0,
    tooWordy: 0,
    tooShort: 0
  };
  
  issues.forEach(issue => {
    issue.issues.forEach(i => {
      if (i.includes('NOT TIME-BOUND')) issueTypes.notTimebound++;
      if (i.includes('NOT SPECIFIC')) issueTypes.notSpecific++;
      if (i.includes('UNCLEAR SUCCESS')) issueTypes.unclearSuccess++;
      if (i.includes('POSSIBLY TOO LONG')) issueTypes.tooLong++;
      if (i.includes('TOO WORDY')) issueTypes.tooWordy++;
      if (i.includes('TOO SHORT')) issueTypes.tooShort++;
    });
  });
  
  console.log(`\nISSUE BREAKDOWN:`);
  console.log(`  Missing time-bound element: ${issueTypes.notTimebound}`);
  console.log(`  Missing specific action: ${issueTypes.notSpecific}`);
  console.log(`  Unclear success criteria: ${issueTypes.unclearSuccess}`);
  console.log(`  Possibly too long (>10 min): ${issueTypes.tooLong}`);
  console.log(`  Too wordy (>100 words): ${issueTypes.tooWordy}`);
  console.log(`  Too short (<30 words): ${issueTypes.tooShort}`);
}

// Generate detailed report file
const reportContent = `# Challenge Actionability Review Report
Generated: ${new Date().toISOString()}

## Summary
- **Total lessons analyzed:** ${lessons.length}
- **Lessons with issues:** ${issues.length}
- **Pass rate:** ${((lessons.length - issues.length) / lessons.length * 100).toFixed(1)}%

## Criteria Checked
1. ✅ **Time-bound:** Challenge mentions "today", "right now", "next time", or specific duration
2. ✅ **Specific action:** Clear verb and concrete task
3. ✅ **Success criteria:** Clear outcome or "notice what happens" prompt
4. ✅ **Quick (<10 min):** Explicitly or implicitly completable in under 10 minutes
5. ✅ **Appropriate length:** 50-75 words (30-100 acceptable)

## Issues Found

${issues.length === 0 ? '🎉 No issues found! All challenges meet actionability criteria.' : ''}

${issues.map(issue => `
### Lesson #${issue.lessonNum}: ${issue.title}
**ID:** ${issue.id}  
**Word count:** ${issue.wordCount}

**Issues:**
${issue.issues.map(i => `- ${i}`).join('\n')}

**Challenge text:**
> ${issue.challenge}

---
`).join('\n')}

## Recommendations

${issues.length > 0 ? `
Based on the analysis, here are the main improvement areas:

1. **Time-bound elements:** ${issues.filter(i => i.issues.some(iss => iss.includes('NOT TIME-BOUND'))).length} challenges need explicit timing (add "right now", "today", "next time", or duration)
2. **Specific actions:** ${issues.filter(i => i.issues.some(iss => iss.includes('NOT SPECIFIC'))).length} challenges need clearer action verbs
3. **Success criteria:** ${issues.filter(i => i.issues.some(iss => iss.includes('UNCLEAR SUCCESS'))).length} challenges need clearer "notice what happens" or measurable outcome
4. **Length issues:** ${issues.filter(i => i.issues.some(iss => iss.includes('WORDY') || iss.includes('SHORT'))).length} challenges need word count adjustment

All issues are fixable with minor edits to improve clarity and actionability.
` : 'All challenges meet the quality standards. No immediate action required.'}
`;

fs.writeFileSync('CHALLENGE_ACTIONABILITY_REPORT.md', reportContent);
console.log(`\n\n📄 Detailed report saved to: CHALLENGE_ACTIONABILITY_REPORT.md`);

File: analyze_challenge_actionability.py
#!/usr/bin/env python3
import re
import json

# Read the SQL file
with open('ascevo/supabase/lessons_all_pillars.sql', 'r', encoding='utf-8') as f:
    sql_content = f.read()

# Split into lesson blocks
lesson_pattern = r"INSERT INTO lessons \(id, unit_id, title, display_order, card_concept, card_example, card_mistake, card_science, card_challenge\) VALUES\s*([\s\S]*?)(?=INSERT INTO lessons|$)"
lesson_blocks = re.findall(lesson_pattern, sql_content)

print(f"Found {len(lesson_blocks)} lesson blocks\n")

# Criteria patterns
immediate_triggers = re.compile(r'\b(Right now|Today|Next time|In your next|This week|Tomorrow|Before|After|When you)\b', re.I)
time_constraint = re.compile(r'\b(\d+\s+(minute|second|hour|day|week)|under \d+|<\d+|less than \d+|for \d+)\b', re.I)
reflection_prompt = re.compile(r'\b(Notice|Pay attention|Observe|Watch|See what|Feel how|Track|Write down|Count|Record)\b', re.I)

issues_found = []
passed_challenges = []
total_lessons = 0

for block in lesson_blocks:
    # Extract individual lessons from the block (can have multiple lessons in one INSERT)
    # Pattern: ('lesson_id','unit_id','title',display_order,'card1','card2','card3','card4','card5'),
    lesson_pattern_individual = r"\('([^']+)','([^']+)','([^']+)',(\d+),\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)',\s*'((?:[^']|'')*)'\)"
    
    lessons_in_block = re.finditer(lesson_pattern_individual, block)
    
    for lesson_match in lessons_in_block:
        lesson_id = lesson_match.group(1)
        unit_id = lesson_match.group(2)
        title = lesson_match.group(3)
        display_order = lesson_match.group(4)
        
        # Cards: concept(5), example(6), mistake(7), science(8), challenge(9)
        challenge_card = lesson_match.group(9).replace("''", "'").strip()
        
        if not challenge_card or len(challenge_card) < 10:
            issues_found.append({
                'lessonId': lesson_id,
                'title': title,
                'issues': ['Missing or empty challenge card'],
                'challengeCard': challenge_card or '(empty)',
                'wordCount': 0
            })
            total_lessons += 1
            continue
        
        total_lessons += 1
        issues = []
        
        # Check criteria
        # 1. Immediate trigger
        if not immediate_triggers.search(challenge_card):
            issues.append('Missing immediate trigger (e.g., "Right now", "Today", "Next time")')
        
        # 2. Time constraint
        if not time_constraint.search(challenge_card):
            issues.append('Missing specific time constraint')
        
        # 3. Reflection prompt
        if not reflection_prompt.search(challenge_card):
            issues.append('Missing reflection prompt (e.g., "Notice what happens")')
        
        # 4. Length check
        if len(challenge_card) > 400:
            issues.append('Challenge may be too long/complex (>400 chars)')
        
        # 5. Word count (target: 50-75 words)
        word_count = len(challenge_card.split())
        if word_count < 30:
            issues.append(f'Too short ({word_count} words, target: 50-75)')
        elif word_count > 100:
            issues.append(f'Too long ({word_count} words, target: 50-75)')
        
        if issues:
            issues_found.append({
                'lessonId': lesson_id,
                'title': title,
                'issues': issues,
                'challengeCard': challenge_card,
                'wordCount': word_count
            })
        else:
            passed_challenges.append({
                'lessonId': lesson_id,
                'title': title,
                'wordCount': word_count
            })

# Generate report
print('=' * 80)
print('CHALLENGE ACTIONABILITY REVIEW REPORT')
print('=' * 80)
print(f'\nTotal Lessons Reviewed: {total_lessons}')
print(f'Challenges Passed: {len(passed_challenges)} ({round(len(passed_challenges)/total_lessons*100)}%)')
print(f'Challenges With Issues: {len(issues_found)} ({round(len(issues_found)/total_lessons*100)}%)')

print('\n' + '=' * 80)
print('ISSUES BY CATEGORY')
print('=' * 80)

# Count issue types
issue_types = {}
for item in issues_found:
    for issue in item['issues']:
        issue_types[issue] = issue_types.get(issue, 0) + 1

for issue, count in sorted(issue_types.items(), key=lambda x: x[1], reverse=True):
    print(f'\n{issue}: {count} lessons')

print('\n' + '=' * 80)
print('DETAILED ISSUES (First 20)')
print('=' * 80)

for idx, item in enumerate(issues_found[:20], 1):
    print(f'\n{idx}. {item["title"]}')
    print(f'   ID: {item["lessonId"]}')
    print(f'   Word Count: {item.get("wordCount", "N/A")}')
    print(f'   Issues:')
    for issue in item['issues']:
        print(f'   - {issue}')
    challenge_preview = item['challengeCard'][:200]
    if len(item['challengeCard']) > 200:
        challenge_preview += '...'
    print(f'   Challenge Card: "{challenge_preview}"')

print('\n' + '=' * 80)
print('SUMMARY & RECOMMENDATIONS')
print('=' * 80)

top_issues = sorted(issue_types.items(), key=lambda x: x[1], reverse=True)[:3]
print(f'''
FINDINGS:
- {len(issues_found)} of {total_lessons} challenges need attention
- Most common issues:''')
for issue, count in top_issues:
    print(f'  • {issue} ({count} challenges)')

print('''
RECOMMENDATIONS:
1. Add immediate triggers to challenges missing them
2. Include specific time constraints (e.g., "5 minutes", "today", "before bed")
3. Add reflection prompts (e.g., "Notice what happens when...", "Track how you feel...")
4. Ensure challenges are completable in under 10 minutes
5. Target 50-75 words per challenge card

NEXT STEPS:
1. Review flagged challenges manually
2. Update SQL file with improved challenge cards
3. Re-run this analysis to verify improvements
''')

# Save detailed report to file
report_data = {
    'summary': {
        'totalLessons': total_lessons,
        'passed': len(passed_challenges),
        'needsWork': len(issues_found),
        'passRate': round(len(passed_challenges)/total_lessons*100)
    },
    'issueTypes': issue_types,
    'issuesFound': issues_found,
    'passedChallenges': passed_challenges
}

with open('challenge_actionability_report.json', 'w', encoding='utf-8') as f:
    json.dump(report_data, f, indent=2, ensure_ascii=False)

print('\nDetailed report saved to: challenge_actionability_report.json')
